src/dft.py

```diff
@@ -4,11 +4,28 @@
 from tree import T
 
 
+#TODO use other structure than lists?
+#TODO don't add leaves?
+#TODO switch not node to node
+
 def in_order(t: T | None) -> Iterable[int]:
     """In-order traversal of a tree.
 
     >>> tree = T(2, T(1, None, None), T(4, T(3, None, None), T(5, None, None)))
     >>> list(in_order(tree))
     [1, 2, 3, 4, 5]
     """
-    return  # FIXME
+    stack = [t]
```

Comment:
I would not do this... Putting `None` on the stack breaks the expected types, I would say. Not that there is anything wrong with having an `Optional` type, but the empty tree has a different nature in my head. But it is mostly a matter of taste, I think.

I also think it makes handling poping empty trees tricky. You need to handle `None` coming from the stack, and that is something you could avoid. It is especially unsatisfactory when you both have to check if you got an empty tree and if the stack is empty.

I'm not saying that this is wrong. I'm just saying that I don't like it. Especially because you also use the placements of `None` on the stack to determine when you have a value and when you have a node, and that is fragile as hell.

I think I would much prefer not to rely on where `None` values are on the stack (and keep them off the stack entirely). Then I can either pop an integer or a node, yield in the first case and push in the second.

```python
def in_order(t: T | None) -> Iterable[int]:
    stack: list[T | int] = []

    def push(x: int | T | None) -> None:
        if x is not None:
            stack.append(x)

    push(t)
    while stack:
        match stack.pop():
            case int(v):
                yield v
            case T(val=val, left=left, right=right):
                push(right)
                push(val)
                push(left)
```